bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java

```diff
@@ -271,28 +275,45 @@ public Row readModifyWriteRow(ReadModifyWriteRowRequest request) {
   private <ReqT, RespT> ListenableFuture<RespT> performAsyncRpc(
       ReqT request, BigtableAsyncRpc<ReqT, RespT> rpc) {
     expandPoolIfNecessary(this.bigtableOptions.getChannelCount());
-    return performRetryingAsyncRpc(request, rpc, null);
+    RetryingUnaryRpcListener<ReqT, RespT> listener =
+        new RetryingUnaryRpcListener<>(
+            retryOptions, request, rpc, retryExecutorService);
+    listener.run();
+    return listener.getCompletionFuture();
   }
 
-  private <ReqT, RespT> RespT performBlockingRpc(ReqT request, BigtableAsyncRpc<ReqT, RespT> rpc) {
-    CancellationToken token = new CancellationToken();
+  private <ReqT, RespT> RespT performBlockingUnaryRpc(ReqT request, BigtableAsyncRpc<ReqT, RespT> rpc) {
+    RetryingUnaryRpcListener<ReqT, RespT> listener =
+        new RetryingUnaryRpcListener<>(retryOptions, request, rpc, retryExecutorService);
     try {
-      return performRetryingAsyncRpc(request, rpc, token).get();
+      listener.run();
+      return listener.getCompletionFuture().get();
     } catch (InterruptedException e) {
-      token.cancel();
+      listener.cancel();
       Thread.currentThread().interrupt();
       throw Status.CANCELLED.withCause(e).asRuntimeException();
     } catch (ExecutionException e) {
-      token.cancel();
+      listener.cancel();
       throw Status.fromThrowable(e).asRuntimeException();
     }
   }
 
-  private <ReqT, RespT> ListenableFuture<RespT> performRetryingAsyncRpc(
-      ReqT request, BigtableAsyncRpc<ReqT, RespT> rpc, CancellationToken cancellationToken) {
-    return new RetryingRpcFunction<>(
-            retryOptions, request, rpc, retryExecutorService, cancellationToken)
-        .callRpcWithRetry();
+  private <ReqT, RespT> List<RespT> performBlockingStreamingRpc(
+      ReqT request, BigtableAsyncRpc<ReqT, RespT> rpc) {
+    RetryingCollectingClientCallListener<ReqT, RespT> listener =
+        new RetryingCollectingClientCallListener<>(
+            retryOptions, request, rpc, retryExecutorService);
+    try {
+      listener.run();
+      return listener.getCompletionFuture().get();
+    } catch (InterruptedException e) {
+      listener.cancel();
+      Thread.currentThread().interrupt();
+      throw Status.CANCELLED.withCause(e).asRuntimeException();
+    } catch (ExecutionException e) {
+      listener.cancel();
+      throw Status.fromThrowable(e).asRuntimeException();
+    }
   }
 
   // Scanner methods
```

Comment:
I'm going to punt on this for now.  We do need to do some refactoring of the scanner logic so that we can also have an observer type interface.  We'll do that another time.